File: 0803-cheapest-flights-within-k-stops/0803-cheapest-flights-within-k-stops.java
class Solution {
    public int findCheapestPrice(int n, int[][] flights, int src, int dst, int k) {
        Map<Integer, Map<Integer, Integer>> graph = new HashMap<>();
        for (int[] flight : flights) {
            graph.putIfAbsent(flight[0], new HashMap<>());
            graph.get(flight[0]).put(flight[1], flight[2]);
        }

        Queue<List<Integer>> pq = new PriorityQueue<>(Comparator.comparingInt(a -> a.get(1)));
        pq.add(Arrays.asList(src, 0, 0));

        Map<Integer, Integer> visited = new HashMap<>();

        while (!pq.isEmpty()) {
            List<Integer> cur = pq.poll();
            int u = cur.get(0);
            int uPrice = cur.get(1);
            int kVisited = cur.get(2);

            if (u == dst) return uPrice;

            visited.put(u, kVisited);

            if (kVisited <= k) {
                kVisited += 1;
                if (graph.containsKey(u)) {
                    for (Map.Entry<Integer, Integer> v : graph.get(u).entrySet()) {
                        if (!visited.containsKey(v.getKey()) || kVisited < visited.get(v.getKey())) {
                            int alt = uPrice + v.getValue();
                            pq.add(Arrays.asList(v.getKey(), alt, kVisited));
                        }
                    }
                }
            }
        }

        return -1;
    }
}

File: 0803-cheapest-flights-within-k-stops/0803-cheapest-flights-within-k-stops.kt
class Solution {
    fun findCheapestPrice(n: Int, flights: Array<IntArray>, src: Int, dst: Int, k: Int): Int {
        val graph: MutableMap<Int, MutableMap<Int, Int>> = mutableMapOf()
        for (flight in flights) {
            graph.putIfAbsent(flight[0], HashMap())
            graph[flight[0]]!![flight[1]] = flight[2]
        }

        val pq: Queue<List<Int>> = PriorityQueue { a: List<Int>, b: List<Int> -> a[1] - b[1] }
        pq.add(listOf(src, 0, 0))

        val visited: MutableMap<Int, Int> = mutableMapOf()

        while (pq.isNotEmpty()) {
            val cur = pq.poll()
            val u = cur[0]
            val uPrice = cur[1]
            var kVisited = cur[2]

            if (u == dst) return uPrice

            visited[u] = kVisited

            if (kVisited <= k) {
                kVisited += 1
                if (graph.containsKey(u)) {
                    for ((key, value) in graph[u]!!) {
                        if (!visited.containsKey(key) || kVisited < visited[key]!!) {
                            val alt = uPrice + value
                            pq.add(listOf(key, alt, kVisited))
                        }
                    }
                }
            }
        }

        return -1
    }
}